### PR TITLE
EES-1132 and EES-1878

### PR DESCRIPTION
**Pull request by markhiveit**
EES-1132 adds extra validation to the release approval checklist. If a release has a generic content section that contain no content blocks, or if a generic content section contains only blank HtmlBlocks, the user will receive a release checklist error.

EES-1878 gives users the ability to attach any methodology to a publication. Previously, the frontend filtered the available methodologies so that only Approved methodologies were available, and the backend validated that created/updated publications methodologies were approved.

### CODE REVIEW

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/explore-education-statistics-admin/src/services/releaseService.ts [hunk 1]
```diff
@@ -116,7 +116,9 @@ export type ReleaseChecklistError =
         | 'DataFileImportsMustBeCompleted'
         | 'DataFileReplacementsMustBeCompleted'
         | 'PublicMetaGuidanceRequired'
-        | 'ReleaseNoteRequired';
+        | 'ReleaseNoteRequired'
+        | 'EmptyContentSectionExists'
+        | 'GenericSectionsContainEmptyHtmlBlock';
     }
   | {
       code: 'MethodologyMustBeApproved';
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/services/releaseService.ts:124]: We should get rid of this as well (remove it from the error mapping code too).

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseChecklistService.cs [hunk 2]
```diff
@@ -101,9 +104,50 @@ public async Task<List<ReleaseChecklistIssue>> GetErrors(Release release)
...
+            return await _contentDbContext.ReleaseContentSections
+                .Include(rcs => rcs.ContentSection)
+                .ThenInclude(cs => cs.Content)
+                .Where(rcs =>
+                    rcs.ReleaseId == releaseId
+                    && rcs.ContentSection.Type == ContentSectionType.Generic)
+                .AnyAsync(rcs => rcs.ContentSection.Content.Count == 0);
+        }
+
+        private bool ReleaseGenericContentSectionsContainEmptyContentBlock(Guid releaseId)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseChecklistService.cs:131]: This method returns `bool` whilst the other returns `Task<bool>`. Would just neaten these up by making them have consistent return types either way.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseChecklistService.cs [hunk 3]
```diff
@@ -101,9 +104,50 @@ public async Task<List<ReleaseChecklistIssue>> GetErrors(Release release)
...
+        {
+            return _contentDbContext.ReleaseContentSections
+                .Include(rcs => rcs.ContentSection)
+                .ThenInclude(cs => cs.Content)
+                .Where(rcs =>
+                    rcs.ReleaseId == releaseId
+                    && rcs.ContentSection.Type == ContentSectionType.Generic)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseChecklistService.cs:138]: Feels like we shouldn't just be checking generic sections. We should probably be checking `Headlines` and `ReleaseSummary` too.

File: src/explore-education-statistics-admin/src/pages/release/components/ReleaseStatusChecklist.tsx [hunk 4]
```diff
@@ -90,6 +90,22 @@ const ReleaseStatusChecklist = ({ checklist, release }: Props) => {
               releaseRouteParams,
             ),
           };
+        case 'EmptyContentSectionExists':
+          return {
+            message: 'Release content sections should not be empty',
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/components/ReleaseStatusChecklist.tsx:95]: Would suggest re-wording to: 'Release content should not contain any empty sections' . This is a little more consistent with the other message and also sounds better for both plural and singular cases.

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseChecklistService.cs (hunk 3 above)
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseChecklistService.cs:138]: Talked to Nick and Cam about this: going create a new ticket to create a separate warning/error for these being empty.

**Review by ntsim — APPROVED** [review]

**Pull request merged**